teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -0,0 +1,205 @@
+---
+status: proposed
+title: Graceful Pipeline Run Termination
+creation-date: '2021-03-18'
+last-updated: '2021-03-18'
+authors:
+- '@@rafalbigaj'
+---
+
+# TEP-0058: Graceful Pipeline Run Termination
+---
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Marking a `PipelineRun` as [cancelled](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#cancelling-a-pipelinerun)
+stops all running tasks and deletes associated Pods. That prevents final tasks, specified in a `Pipeline` under `finally` section,
+from being run.
+There should be a way to terminate a `PipelineRun` gracefully and wait for cleanup actions triggered by final tasks.
```

Comment:
Thanks @rafalbigaj, this is a reasonable summary and ask.

What would `gracefully` mean here? With the `PipelineRunCancel`, the existing `taskRuns` are exited with  `Cancelled(TaskRunCancelled)` and pipelineRun exits with `Cancelled(PipelineRunCancelled)`. What is your expectation in terms of the exit status of existing `taskRuns`?

Reply:
@pritidesai This is a good question. I can think about 2 use cases and probably your next comment could somehow address the second one.

1. Stop running tasks and mark them as cancelled
2. Wait for running tasks to finish

... and then cancel a pipeline run.

I have added the requirement to wait for running tasks in the "Requirements" section.